File: ichrisbirch/services/row_limit.py
"""Row capping for the collection reads, with one meaning for zero.

`cli-design.md` § "`--follow`/`-f` defaults to false; `--limit`/`-n` goes on
every list" puts a limit on every list over data that grows outside the binary,
which is every collection this API answers. Capping here rather than in each
client is `data.md` § "Filtering is server-side": the rows the caller discards
are rows nobody had to serialize.

**Absence means no cap and zero means no rows.** `cli-design.md` § "A sentinel
never steals a value the caller can mean" is the rule: a reserved value has to
be one no caller could have meant, and `tail -n 0` settles that a caller can
mean zero rows. So `limit` omitted leaves the query uncapped, and `limit=0`
reaches `LIMIT 0` and returns an empty set. One helper is what keeps every
endpoint on the same side of that.

A negative limit is rejected at the edge by `ge=0` rather than reaching SQL.
"""

from typing import Annotated

from fastapi import Query
from sqlalchemy import Select

RowLimit = Annotated[
    int | None,
    Query(ge=0, description='Return at most this many rows. Omit for every row; 0 returns none.'),
]


def apply_row_limit(query: Select, limit: int | None) -> Select:
    """Cap the query at `limit` rows, or leave it alone when there is no cap.

    `None` is the absence of a cap and is the only thing that asks for every
    row. `0` is a row count like any other, so it reaches `LIMIT 0`. A falsy
    test cannot tell the two apart, and it answers the caller that asked for
    nothing with the whole collection.
    """
    if limit is None:
        return query
    return query.limit(limit)

File: cli/internal/api/row_limit.go
package api

import (
	"net/url"
	"strconv"
)

// applyLimit writes a row cap into params, or writes nothing when there is no
// cap to apply. One helper across every list read because --limit means the
// same thing on each: at most this many rows, server-side.
//
// A nil limit is the absence of a cap and is not sent, which is what asks for
// every row. A limit of zero is a row count the caller can mean, so it is sent
// as `limit=0` and the API answers with nothing. Folding the two together
// answers a caller computing its own bound with the whole collection.
func applyLimit(params url.Values, limit *int) {
	if limit == nil {
		return
	}
	params.Set("limit", strconv.Itoa(*limit))
}
